pallets/remote-asset-manager/src/calls/staking.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2021 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+//! Xcm support for `pallet_staking` calls
+
+use codec::{Compact, Decode, Encode, Output};
+use frame_support::{
+    sp_std::vec::Vec,
+    weights::{constants::RocksDbWeight, Weight},
+    RuntimeDebug,
+};
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+use crate::calls::PalletCallEncoder;
+use crate::{CallEncoder, EncodeWith, PalletCall};
+
+/// The index of `pallet_staking` in the polkadot runtime
+pub const POLKADOT_PALLET_STAKING_INDEX: u8 = 7u8;
+
+/// Provides encoder types to encode the associated types of the  `pallet_staking::Config` trait depending on the configured Context.
+pub trait StakingCallEncoder<Source, Balance, AccountId>: PalletCallEncoder {
+    /// Encodes the `<pallet_staking::Config>::Balance` depending on the context
+    type CompactBalanceEncoder: EncodeWith<Balance, Self::Context>;
+
+    /// Encodes the `<pallet_staking::Config>::Source` depending on the context
+    type SourceEncoder: EncodeWith<Source, Self::Context>;
+
+    /// Encodes the `<pallet_staking::Config>::AccountId` depending on the context
+    type AccountIdEncoder: EncodeWith<AccountId, Self::Context>;
+}
+
+impl<'a, 'b, Source, Balance, AccountId, Config> Encode
+    for CallEncoder<'a, 'b, StakingCall<Source, Balance, AccountId>, Config>
+where
+    Config: StakingCallEncoder<Source, Balance, AccountId>,
+{
+    fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+        // include the pallet identifier
+        dest.push_byte(self.call.pallet_call_index());
+        match self.call {
+            StakingCall::Bond(bond) => {
+                Config::SourceEncoder::encode_to_with(&bond.controller, self.ctx, dest);
+                Config::CompactBalanceEncoder::encode_to_with(&bond.value, self.ctx, dest);
+
+                match &bond.payee {
+                    RewardDestination::Staked => {
+                        dest.push_byte(0);
+                    }
+                    RewardDestination::Stash => {
+                        dest.push_byte(1);
+                    }
+                    RewardDestination::Controller => {
+                        dest.push_byte(2);
+                    }
+                    RewardDestination::Account(ref account) => {
+                        dest.push_byte(3);
+                        Config::AccountIdEncoder::encode_to_with(account, self.ctx, dest);
+                    }
+                    _ => {}
```

Comment:
Any table or document for this encode?

Reply:
That's how `scale` codec expands to for enums this is copy pasted from the staking pallet, see also
[pallet-staking-expanded.rs.txt](https://github.com/ChainSafe/PINT/files/6609548/pallet-staking-expanded.rs.txt)